File: LeetCode 热题 HOT 100/栈和队列/232. 用栈实现队列.py
# 请你仅使用两个栈实现先入先出队列。队列应当支持一般队列支持的所有操作（push、pop、peek、empty）：

# 实现 MyQueue 类：

# void push(int x) 将元素 x 推到队列的末尾
# int pop() 从队列的开头移除并返回元素
# int peek() 返回队列开头的元素
# boolean empty() 如果队列为空，返回 true ；否则，返回 false

# 说明：
# 你 只能 使用标准的栈操作 —— 也就是只有 push to top, peek/pop from top, size, 和 is empty 操作是合法的。
# 你所使用的语言也许不支持栈。你可以使用 list 或者 deque（双端队列）来模拟一个栈，只要是标准的栈操作即可。


# 示例 1：
# 输入：
# ["MyQueue", "push", "push", "peek", "pop", "empty"]
# [[], [1], [2], [], [], []]
# 输出：
# [null, null, null, 1, 1, false]
# 解释：
# MyQueue myQueue = new MyQueue();
# myQueue.push(1); // queue is: [1]
# myQueue.push(2); // queue is: [1, 2] (leftmost is front of the queue)
# myQueue.peek(); // return 1
# myQueue.pop(); // return 1, queue is [2]
# myQueue.empty(); // return false


class MyQueue:
  
    def __init__(self):
      self.in_stack = list()
      self.out_stack = list()

    def push(self, x):
      self.in_stack.append(x)
        

    def pop(self):
      if self.in_stack != []:
        for i in range(len(self.in_stack) - 1):
          temp = self.in_stack.pop()
          self.out_stack.append(temp)
        res = self.in_stack.pop()
        for i in range(len(self.out_stack)):
          temp = self.out_stack.pop()
          self.in_stack.append(temp)
        return res

    def peek(self):
      if self.in_stack != []:
        for i in range(len(self.in_stack) - 1):
          temp = self.in_stack.pop()
          self.out_stack.append(temp)
        res = self.in_stack.pop()
        self.out_stack.append(res)
        for i in range(len(self.out_stack)):
          temp = self.out_stack.pop()
          self.in_stack.append(temp)
        return res

    def empty(self):
      return self.in_stack == []

File: LeetCode 热题 HOT 100/栈和队列/232. 用栈实现队列.js
var MyQueue = function () {
  this.in_stack = []
  this.out_stack = []
};

/**
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function (x) {
  this.in_stack.push(x)
};

/**
 * @return {number}
 */
MyQueue.prototype.pop = function () {
  if (!this.outStack.length) {
    this.in2out();
  }
  return this.outStack.pop();
};

/**
 * @return {number}
 */
MyQueue.prototype.peek = function () {
  if (!this.outStack.length) {
    this.in2out();
  }
  return this.outStack[this.outStack.length - 1];
};

/**
 * @return {boolean}
 */
MyQueue.prototype.empty = function () {
  return this.outStack.length === 0 && this.inStack.length === 0;
};

MyQueue.prototype.in2out = function () {
  while (this.inStack.length) {
    this.outStack.push(this.inStack.pop());
  }
}